### PR TITLE
`polkadot-parachain`: add manual seal support

### PR DESCRIPTION
**Pull request by serban300**
Resolves https://github.com/paritytech/polkadot-sdk/issues/5026

This PR adds support for starting a dev node with manual seal consensus. This can be done by using the `--dev-block-time` argument . For example:
```
polkadot-parachain --chain asset-hub-rococo-dev --dev-block-time 5000 --tmp
```

### CODE REVIEW

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:215]: I am a bit unsure as to why this limitation should exist. So this will only work if the chain-spec's id ends with -dev. Why should this limitation exist?

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs:126]: The raw manual seal also requires you to submit some kind of an RPC to author a block. What I have done in the past, I think this is a good improvement to it, is to run an RPC client internally that pings the manual seal server and produces blocks every x ms. 

This is then how I re-formulated this CLI argument as `--dev-block-time 100`, which produces a block every 100ms.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:215]: I will investigate and get back with the details, but the general problem is that I'm getting errors at block production when trying to use a non-dev runtime (for example: `asset-hub-polkadot`).

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (skunert) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:307]: If we want to get rid of this, you could try to add
```rust
.instrument(sc_tracing::tracing::info_span!(
sc_tracing::logging::PREFIX_LOG_SPAN,
name = "Parachain",
)))
```
to the futures that are being run. This is what this macro should also be doing.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (skunert) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:224]: Did not check this in-depth, but could we not mingle the dev_block_time into extra args and make the new_node_spec above have something like `Consensus::Manual`. I am wondering if we can get rid of the "special handling" of `start_manual_seal` and just treat it as a consensus variant.

**Review by skunert — COMMENTED** [review]
Did not look at the `manual_seal.rs` code itself yet, will make one more pass.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:224]: I don't think we can treat the manual seal just like another consensus variant. Because the process of starting the node is a bit different (the relay_chain_interface is not built), and the arguments passed to `start_consensus` are also different. Mixing this in the `NodeSpec` feels a bit dirty.

That's why I just added a new type of node with a new `start_manual_seal_node()` fn that differs from `start_node()`. We could probably make it more generic and name it `SoloChainNode` instead of `ManualSealNode`, if it makes any difference, but I would keep it separate.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs
Comment (skunert) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs:63]: Nit: We could implement BuildImportQueue directly for `ManualSealNode` instead of having the extra struct. Just a suggestion, do as you like.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs
Comment (skunert) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs:134]: Nit: Could be essential, if for whatever reason sending to the sink fails we can crash.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs
Comment (skunert) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs:110]: We need to make sure here that network does not try to sync. Maybe setting the in/out peers to 0 or maybe we have a mock somewhere. The dev node does not really need networking.

Problems occured when I was playing with this and invoked for example `polkadot-parachain --dev-block-time 3000 --tmp --chain asset-hub-rococo.json`. The node found some peers and tried to import blocks from them, which failed.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (skunert) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:224]: Fair enough, my thought was that the node starting of the manual seal basically has different args, but its a subset of the usual ones. So we could have some kind of wrapper around the node were start_node internally invokes manual seal variant if extra_args contains the dev node arg. 

But not important, we can leave it as is.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs]: Good point ! Done.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/manual_seal.rs:107]: Good point ! Done.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:224]: Thanks ! Will leave it as it is then.

**Review by skunert — APPROVED** [review]
Nice!

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/mod.rs
Comment (skunert) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/mod.rs:50]: Just FYI, I took a shot at removing this workaround, should merge soon: https://github.com/paritytech/polkadot-sdk/pull/5839

Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/mod.rs:50]: Nice ! Thanks !

Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/mod.rs]: Merged your fix in the PR branch and removed `UnwindSafe` and `RefUnwindSafe`

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/shell.rs
Comment (kianenigma) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/shell.rs:1]: Yeah it is not needed, but it can be an easy follow-up. If not done here, please make an issue and add it to our project board please :) 

It can also be a good one to have the mentor tag and use externals. Seems very straightforward.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/mod.rs
Comment (kianenigma) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/mod.rs:26]: Some docs about why it exists and its intended purpose would help.

**Review by kianenigma — APPROVED** [review]
LGTM assuming all open discussions are well addressed and closed :)

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/shell.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/shell.rs:1]: Opened a new issue for this since it seems a bit separate: https://github.com/paritytech/polkadot-sdk/issues/5898 . I assigned it to myself and I plan to do it immediately after merging this one.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs]: Good point ! Done.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/mod.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/nodes/mod.rs:30]: Done.

**Comment by kianenigma** [comment]
Final request: PR description also seems out of date wrt example 🙏

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs:128]: I now wonder if the difference between `--dev-block-time` and `--dev` is clear? 

Perhaps nitpicking, and ofc this can be fixed with great documentation.

cc @skunert

Comment (kianenigma) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs:128]: First glance, what comes to mind is a subcommand here that doesn't inherit all the other `SharedParams` and such, but I think that has already been explored?

**Comment by serban300** [comment]
> Final request: PR description also seems out of date wrt example 🙏

Sorry, missed it. Done.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs:128]: Yes, a separate subcommand was explored before, but I admit I didn't think about this issue back then. But even if we add a separate subcommand, from what I understand it still needs to contain the `SharedParams` in order to be able to create the runner: 

```
fn create_runner<T: CliConfiguration<DVC>, DVC: DefaultConfigurationValues>(
		&self,
		command: &T,
	) -> Result<Runner<Self>>

...

pub trait CliConfiguration<DCV: DefaultConfigurationValues = ()>: Sized {
	/// Get the SharedParams for this object
	fn shared_params(&self) -> &SharedParams;
       
       ...
}
```

Anyway, it is a good point. But could we explore it in a separate issue ? It would be good if we could merge this one before doing the renamings in https://github.com/paritytech/polkadot-sdk/issues/5566 and in general before doing other changes to `polkadot-parachain-lib/bin` in order to avoid merging conflicts.

Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/cli.rs:128]: Opened separate issue for this as discussed offline: https://github.com/paritytech/polkadot-sdk/issues/5903

**Pull request merged**
